package/MDAnalysis/analysis/dssp/dssp.py

```diff
@@ -0,0 +1,337 @@
+
+    Raises
+    ------
+    ValueError
+        if `guess_hydrogens` is True but some non-PRO hydrogens are missing.
+
+    Examples
+    --------
+
+    For example, you can assign secondary structure for a single PDB file:
+
+    .. code-block:: python
+        from MDAnalysis.analysis.dssp import DSSP
+        from MDAnalysisTests.datafiles import PDB
+        import MDAnalysis as mda
+        u = mda.Universe(PDB)
+        run = DSSP(u).run()
+        print("".join(run.results.dssp[0][:20]))
+        # '--EEEEE-----HHHHHHHH'
+
+    Also, per-frame dssp assignment allows you to build average
+    secondary structure -- `DSSP.results.dssp_ndarray` holds
+    (n_frames, n_residues, 3) shape ndarray with one-hot encoding of
+    loop, helix and sheet, respectively:
+
+    .. code-block:: python
+        from MDAnalysis.analysis.dssp import translate
+        u = mda.Universe(...)
+        long_run = DSSP(u).run()
+        mean_secondary_structure = translate(
+            long_run.results.dssp_ndarray.mean(axis=0)
+            )
+        print(''.join(mean_secondary_structure)[:20])
+        # '---HHHHHHHH---------'
+
+    .. versionadded:: 2.8.0
+    """
+
+    def __init__(self, u: Universe, guess_hydrogens: bool = True):
+        super().__init__(u.trajectory)
+        self._guess_hydrogens = guess_hydrogens
+
+        # define necessary selections
+        # O1 is for C-terminal residue
+        heavyatom_names = ("N", "CA", "C", "O O1")
+        self._heavy_atoms: dict[str, "AtomGroup"] = {
+            t: u.select_atoms(f"protein and name {t}") for t in heavyatom_names
+        }
+        self._hydrogens: list["AtomGroup"] = [
+            res.atoms.select_atoms("name H")
+            for res in u.select_atoms("protein").residues
+        ]
+        # can't do it the other way because I need missing values to exist
+        # so that I could fill them in later
+
+        if not self._guess_hydrogens:
+            for calpha, hydrogen in zip(
+                self._heavy_atoms["CA"][1:], self._hydrogens[1:]
+            ):
+                if not hydrogen and calpha.resname != "PRO":
+                    raise ValueError(
+                        (
+                            "Universe is missing non-PRO hydrogen "
+                            f"on residue {calpha.residue}"
+                        )
+                    )
+
+    def _prepare(self):
+        self.results.dssp_ndarray = []
+
+    def _get_coords(self) -> np.ndarray:
+        """Returns coordinates of (N,CA,C,O,H) atoms, as required by
+        :func:`get_hbond_map` and :func:`assign` functions.
+
+        Returns
+        -------
+        np.ndarray
+            coordinates of (N,CA,C,O,H) atoms
+
+        Raises
+        ------
+        ValueError
+            if input Universe contains different number of (N,CA,C,O) atoms
+
+        .. versionadded:: 2.8.0
+        """
+        positions = [group.positions for group in self._heavy_atoms.values()]
```

Comment:
The above will not work because when selecting O we omit the OT1,2 at the C-terminus and thus not all residues have the same number of atoms per residue.

For what it's worth, one could also create a dict with
```python
In [45]: u.select_atoms("name N CA C O").groupby("names")
Out[45]:
{'CA': <AtomGroup with 214 atoms>,
 'C': <AtomGroup with 214 atoms>,
 'O': <AtomGroup with 213 atoms>,
 'N': <AtomGroup with 214 atoms>}
```

Reply:
>For what it's worth, one could also create a dict with

I'll give it a spin and benchmark on some trajectory (after I fix the checks above), and simply add the one that is faster.

Reply:
>For what it's worth, one could also create a dict with

But we need both O & O1 atoms, and I couldn't figure out a way to make it work together in this `groupby` approach.

>One could make it faster by having all the heavy atoms in a single AtomGroup and then reshaping positions:
Turns out it's not the bottleneck, and doesn't make things any faster (in my hands at least) -- the difference between both approaches is within standard deviation in 5 attempts on gromacs lysozyme in water, so I'll leave it as is.

Reply:
thanks for checking